app/models/concerns/post_concerns/autoflagging.rb

```diff
@@ -19,8 +19,11 @@ def autoflag
       begin
         conditions = post.site.flag_conditions.where(flags_enabled: true)
         available_user_ids = {}
+        weight = post.reasons.pluck(:weight).reduce(:+)
+        poster_rep = post.stack_exchange_user.reputation
+        reasons_count = post.reasons.count
         conditions.each do |condition|
-          if condition.validate!(post)
+          if condition.validate_weight_rep_reason_count!(weight, poster_rep, reasons_count)
```

Comment:
Why aren't you using `condition.validate_post! post` here?

Reply:
The point of this PR is to not use that function there. I should have quoted the [chat conversation](https://chat.stackexchange.com/transcript/11540?m=54849478#54849478) in the PR, rather than just linking to it. I'm sorry I didn't put that context directly in the PR.

Basically, `condition.validate!(post)` calls `post.reasons.pluck(:weight).reduce(:+)` every time `condition.validate!(post)` is called (i.e. each iteration through the loop). That's the way that function needs to handle it in order to validate an arbitrary post. That's good and modular OOP. However, it's inefficient when wanting to validate the same post against multiple flagging conditions.

In the loops this PR changes, we're looping through all enabled flagging conditions to validate each condition against the same post. We know we're testing the same post every loop iteration, just against a different flagging condition. So, the value returned by `post.reasons.pluck(:weight).reduce(:+)` will be the same through every iteration of the loop, but the call to `condition.validate!(post)` still results in a database access in each loop iteration to get a value which won't change.

There are, currently, 450+ existing flagging conditions.  Of those, 330+ are enabled. So, the existing code makes 330+ redundant database accesses to get the post's weight when traversing this loop, which is done for every post reported to MS. The purpose of this PR is to pull that database access out of the loop, so it's only done once, prior to the loop. The value resulting from that database access is then used in each loop iteration.

This PR should save us a few to several hundred database accesses for each post reported to MS.

Reply:
Oh yes, I should've understood this sooner.

Reply:
If I were to work on this, I would have changed `Post` to add caches for these values, maybe something like this:

```ruby
class Post
  def weight
    @weight ||= reasons.pluck(:weight).reduce(:+)
  end

  def poster_rep
    @poster_rep ||= stack_exchange_user.reputation
  end

  def reason_count
    @reason_count ||= reasons.count
  end
end 
```

Then stick to `validate! post`, which is cleaner and more Ruby-ish IMO.

Reply:
That's a reasonable way to do it. In fact, doing basically that was actually the first thing I thought of. I just don't know Ruby well enough, or the data structures used in MS, to be confident implementing it in that manner without *actually* testing, and I don't have an MS development environment set up in order to do that testing.

Reply:
If you'd like to implement it that way and submit a new PR, I'm perfectly OK with that being the way this is resolved.

Reply:
Here's my idea summed up as code:

<https://github.com/iBug/metasmoke/commit/89fb512197e90f743f0e3839f93dc79e6d65fb50>